### PR TITLE
GET/ attachment by id

### PR DESCRIPTION
**Pull request by nitro-marky**
**Problem**
Need a GET endpoint to get attachment by id

**Solution**
Implement endpoint through to db method, API route and API Service.

### CODE REVIEW

File: app/api-v1/routes/attachment/{id}.js [hunk 1]
```diff
@@ -1,8 +1,26 @@
 // eslint-disable-next-line no-unused-vars
+const { BadRequestError } = require('../../../utils/errors')
 module.exports = function (attachmentService) {
   const doc = {
     GET: async function (req, res) {
-      res.status(500).json({ message: 'Not Implemented' })
+      const response = await attachmentService.getAttachmentByID(req.params.id)
+
+      if (req.headers.accept.includes('json') && response.filename === 'json') {
```
Comment (mattdean-digicatapult) [inline app/api-v1/routes/attachment/{id}.js:8]: I think we need to do a more correct parsing of the accept header here. I'm not sure if there's an existing module (hopefully) but the standard is well documented https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Accept. Essentially if the file is JSON and the accept header prefers JSON we should return that. If the file is not json but there is an accept entry for `application/octet-stream`, `application/*` or `*/*` we should return the octet stream. We might need to work out some test cases to enumerate the different behaviours we want then use that to test our implmentation

File: app/api-v1/services/attachmentService.js [hunk 2]
```diff
@@ -17,7 +17,14 @@ const createAttachment = async (name, buffer) => {
   return attachment
 }
 
+async function getAttachmentByID(id) {
+  const [attachmentResult] = await getAttachmentByIdDb(id)
+
+  return attachmentResult
+}
```
Comment (mattdean-digicatapult) [inline app/api-v1/services/attachmentService.js:24]: need to handle if this does not exist

File: test/integration/attachment.test.js [hunk 3]
```diff
@@ -76,4 +84,17 @@ describe('attachments', function () {
     expect(response.status).to.equal(400)
     expect(response.error.text).to.equal('Unexpected token t in JSON at position 0')
   })
+
+  it('should return from the JSON route', async function () {
```
Comment (mattdean-digicatapult) [inline test/integration/attachment.test.js:88]: Also need error case for if an `application/json` response is requested but the file is not `json`

File: app/api-v1/routes/attachment/{id}.js [hunk 4]
```diff
@@ -1,8 +1,26 @@
 // eslint-disable-next-line no-unused-vars
+const { BadRequestError } = require('../../../utils/errors')
 module.exports = function (attachmentService) {
   const doc = {
     GET: async function (req, res) {
-      res.status(500).json({ message: 'Not Implemented' })
+      const response = await attachmentService.getAttachmentByID(req.params.id)
```
Comment (jonmattgray) [inline app/api-v1/routes/attachment/{id}.js:6]: test for invalid id?

File: app/api-v1/routes/attachment/{id}.js [hunk 5]
```diff
@@ -1,8 +1,26 @@
...
+
+      if (req.headers.accept.includes('json') && response.filename === 'json') {
+        const json = JSON.parse(response.binary_blob)
```
Comment (jonmattgray) [inline app/api-v1/routes/attachment/{id}.js:9]: could `JSON.parse` fail?

**Review by mattdean-digicatapult — COMMENTED** [review]
Some specific comments on your implementation so fair. In short I think we need to build up some examples

File: app/api-v1/routes/attachment/{id}.js [hunk 6]
```diff
@@ -1,8 +1,26 @@
...
+        res.status(200).json(json)
+      } else if (req.headers.accept.includes('json') && response.filename != 'json') {
+        throw new BadRequestError({ message: 'Filename does not match JSON' })
```
Comment (jonmattgray) [inline app/api-v1/routes/attachment/{id}.js:12]: I reckon saying something along lines of `not a JSON file` would be clearer to the user than talking about filename as renaming the file to `json` happens in code

**Comment by mattdean-digicatapult** [comment]
There is also the question of what to do if the `Accept` header doesn't contain something that allows `application/octet-stream` but the repsonse must be `application/octet-stream` because of the file type. In that case we can either return [`406`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/406) or we can return the `application/octet-stream` anyway.  Happy to discuss

File: app/api-v1/routes/attachment/{id}.js (hunk 1 above)
Comment (nitro-marky) [reply app/api-v1/routes/attachment/{id}.js:8]: At the moment anything that isn't `application/json` will return an `octect-stream` by default.

Comment (nitro-marky) [reply app/api-v1/routes/attachment/{id}.js:8]: At the moment it defaults to an `octet-stream` so do you want individual accept type checking?

File: app/api-v1/routes/attachment/{id}.js [hunk 7]
```diff
@@ -1,8 +1,61 @@
 // eslint-disable-next-line no-unused-vars
+const { BadRequestError, InternalError, NotAcceptableError } = require('../../../utils/errors')
+const { parseAccept } = require('../../../utils/parseAcceptHeader')
+
 module.exports = function (attachmentService) {
   const doc = {
     GET: async function (req, res) {
-      res.status(500).json({ message: 'Not Implemented' })
+      const { validate: uuidValidate } = require('uuid')
```
Comment (mattdean-digicatapult) [inline app/api-v1/routes/attachment/{id}.js:8]: `require` should always be at the top of the file

File: app/api-v1/routes/attachment/{id}.js [hunk 8]
```diff
@@ -1,8 +1,61 @@
...
+      if (!uuidValidate(req.params.id)) {
+        throw new BadRequestError({ message: 'Id not valid uuid format' })
+      }
```
Comment (mattdean-digicatapult) [inline app/api-v1/routes/attachment/{id}.js:11]: do we need this check? Is the parameter not validated automatically?

File: helm/inteli-api/Chart.yaml [hunk 9]
```diff
@@ -1,8 +1,8 @@
 apiVersion: v2
 name: inteli-api
-appVersion: '1.20.3'
+appVersion: '1.23.0'
```
Comment (mattdean-digicatapult) [inline helm/inteli-api/Chart.yaml:3]: why the big version bump?

File: test/helper/routeHelper.js [hunk 10]
```diff
@@ -140,6 +140,38 @@ async function postAttachmentJSON({ app }, body, token) {
...
+      console.error(`postAttachmentErr ${err}`)
+      return err
+    })
+}
+
+async function getAttachmentRouteOctet(id, { app }, token) {
+  return request(app)
+    .get(`/${API_MAJOR_VERSION}/attachment/${id}`)
+    .set('Accept', 'application/octet-stream')
+    .set('Content-Type', 'application/octet-stream')
```
Comment (mattdean-digicatapult) [inline test/helper/routeHelper.js:163]: No need for a `Content-Type` header, this is a get

File: test/helper/routeHelper.js [hunk 11]
```diff
@@ -140,6 +140,38 @@ async function postAttachmentJSON({ app }, body, token) {
     })
 }
 
+async function getAttachmentRouteJSON(id, { app }, token) {
+  return request(app)
+    .get(`/${API_MAJOR_VERSION}/attachment/${id}`)
+    .set('Accept', 'application/json')
+    .set('Content-Type', 'application/json')
```
Comment (mattdean-digicatapult) [inline test/helper/routeHelper.js:147]: No need for a `Content-Type` header, this is a get

**Review by mattdean-digicatapult — CHANGES_REQUESTED** [review]
Really really close @nitro-marky! Just a couple really minor tweaks

File: app/api-v1/routes/attachment/{id}.js (hunk 8 above)
Comment (nitro-marky) [reply app/api-v1/routes/attachment/{id}.js:11]: Jonathan suggested validating IDs 🤷

File: helm/inteli-api/Chart.yaml (hunk 9 above)
Comment (nitro-marky) [reply helm/inteli-api/Chart.yaml:3]: I was on 22 then a conflict merge appeared which all the way back down. I'll try and make this mystery number game happen at a lower level.

File: app/api-v1/routes/attachment/{id}.js (hunk 8 above)
Comment (mattdean-digicatapult) [reply app/api-v1/routes/attachment/{id}.js:11]: if you remove it does a test fail?

Comment (jonmattgray) [reply app/api-v1/routes/attachment/{id}.js:11]: sorry by test for invalid I meant, test what happens if no attachment exists for that `id`

Comment (jonmattgray) [reply app/api-v1/routes/attachment/{id}.js:11]: invalid was the wrong word to use, my bad

Comment (mattdean-digicatapult) [reply app/api-v1/routes/attachment/{id}.js:11]: yeah we should have a `404` case. Also we should handle invalid `uuid` values. What I'm questioning is whether this is done for us by the openapi framework or if we have to do it like this ourselves

**Review by mattdean-digicatapult — APPROVED** [review]

**Review by jonmattgray — APPROVED** [review]

**Pull request merged**